Index: asterisk-17.4.0/res/res_rtp_asterisk.c
===================================================================
--- asterisk-17.4.0.orig/res/res_rtp_asterisk.c
+++ asterisk-17.4.0/res/res_rtp_asterisk.c
@@ -3578,6 +3578,7 @@ static void rtp_add_candidates_to_ice(st
 					pj_sockaddr_get_len(&pjtmp));
 			++count;
 		}
+		freeifaddrs(ifa);
 	}
 
 	/* If configured to use a STUN server to get our external mapped address do so */
